src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs

```diff
@@ -325,6 +346,53 @@ public Task<IAsyncCursor<RawBsonDocument>> ResumeAsync(IReadBinding binding, Can
             return aggregateOperation.ExecuteAsync(binding, cancellationToken);
         }
 
+        /// <inheritdoc />
+        public void ReviewResumeParams()
+        {
+            // Docs: if the ChangeStream has a cached postBatchResumeToken and has returned all documents in the most recent batch(or the most recent batch was empty):
+            // The driver MUST set resumeAfter to the cached postBatchResumeToken.
+            // The driver MUST NOT set startAfter. If startAfter was in the original aggregation command, the driver MUST remove it.
+            // The driver MUST NOT set startAtOperationTime. If startAtOperationTime was in the original aggregation command, the driver MUST remove it.
+            // todo: should be implemented in scope of the CSHARP-2514.
+
+            // Docs: if the ChangeStream has a cached documentResumeToken:
+            // The driver MUST set resumeAfter to the cached documentResumeToken.
+            // The driver MUST NOT set startAfter. If startAfter was in the original aggregation command, the driver MUST remove it.
+            // The driver MUST NOT set startAtOperationTime. If startAtOperationTime was in the original aggregation command, the driver MUST remove it.
+            // todo:
+
+            // Docs: if the ChangeStream was created with startAfter:
+            // The driver MUST set resumeAfter to the value of the originally used startAfter.
+            // The driver MUST NOT set startAfter. The driver MUST remove it from the original aggregation command.
+            // The driver MUST NOT set startAtOperationTime.
+            if (_startAfter != null)
+            {
+                _resumeAfter = _startAfter;
+                _startAfter = null;
+                _startAtOperationTime = null;
+            }
+            // Docs: if the ChangeStream was created with resumeAfter:
+            // The driver MUST set resumeAfter to the value of the originally used resumeAfter.
+            // The driver MUST NOT set startAfter.
+            // The driver MUST NOT set startAtOperationTime.
+            else if (_resumeAfter != null)
+            {
+                _startAfter = null;
+                _startAtOperationTime = null;
+            }
+
+            // Docs: if the ChangeStream has a saved operation time (either from an originally
```

Comment:
We should send the correct values for resumeAfter, startAfter and startAtOperationTime to the server.

Perhaps it would be easier to also implement support for postBatchResumeToken at the same time in order to not have to figure out how to implement "half" of the resume logic.